rtos/MessagePipe.h

```diff
@@ -0,0 +1,264 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2017 ARM Limited
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#ifndef MESSAGE_PIPE_H
+#define MESSAGE_PIPE_H
+
+#include <stdint.h>
+#include <errno.h>
+
+#include "cmsis_os2.h"
+#include "MemoryPool.h"
+#include "EventFlags.h"
+
+#include "platform/platform.h"
+#include "platform/mbed_critical.h"
+#include "platform/FileHandle.h"
+#include "platform/NonCopyable.h"
+
+namespace rtos {
+/** \addtogroup rtos */
+/** @{*/
+
+/** The MessagePipe class is abstraction of FileHandle over Mail(subsequently Queue).
+ A message is a memory block that is send to a thread or interrupt service routine.
+ The MessagePipe should be configured to be not blocking when used in interrupt service routine to prevent blocking calls.
+
+  @tparam  T         data type of a single message element.
+  @tparam  queue_sz  maximum number of messages in pipe.
+
+*/
+
+template<typename T, uint32_t pipe_sz>
+class MessagePipe : public mbed::FileHandle, private mbed::NonCopyable<MessagePipe<T, pipe_sz> > {
+
+public:
+    MessagePipe()
+    {
+        set_blocking(true);
+    }
+
+    /** Equivalent to POSIX poll(). Derived from FileHandle.
+     *  Provides a mechanism to multiplex input/output over a set of file handles.
+     */
+    virtual short poll(short events) const {
+        short revents = 0;
+
+        if (!_mail.empty() && (events & POLLIN)) {
+            revents |= POLLIN;
+        }
+
+        if (!_mail.full() && (events & POLLOUT)) {
+            revents |= POLLOUT;
+        }
+
+        return revents;
+    }
+
+    /** Write the contents of a buffer to a file
+     *
+     *  @param buffer   The buffer to write from
+     *  @param length   The number of bytes to write
+     *  @return         The number of bytes written, negative error on failure
+     */
+    virtual ssize_t write(const void* buffer, size_t length) {
```

Comment:
Would it be desirable to handle the case where multiple objects have to be written in a single transaction ?